Feat/custom fields TinaCMS docs (#1024)
tinacms/tinacms.org: content/docs/advanced-features/custom-fields.md

---
title: Custom Fields
prev: Null
next: null
---

Custom fields are made possible by the [plugin system](/docs/plugins/) in TinaCMS. Custom fields are a way to either add [custom logic to existing fields](#adding-custom-logic) or provide an [entirely new field component](#adding-a-custom-field)

## Adding Custom Logic

> Adding custom logic is easy and can be done in two simple steps; registering a field plugin and then using the field plugin in your schema.

Reviewer: ```suggestion
Adding custom logic involves two steps: first registering a field plugin and then using the field plugin in your schema.
```

Avoid using easy, simple, just, etc. It doesn't help anyone and is not inclusive.

### 1. Create a Field Plugin

**Interface**
```ts
export interface FieldPlugin<ExtraFieldProps = {}, InputProps = {}> {
  __type: 'field'
  name: string
  Component: React.FC<InputFieldType<ExtraFieldProps, InputProps>>
  type?: string
  validate?(
    value: any,
    allValues: any,
    meta: any,
    field: Field
  ): string | object | undefined
  parse?: (value: any, name: string, field: Field) => any
  format?: (value: any, name: string, field: Field) => any
  defaultValue?: any
}

```

Where component can be a [custom component](#adding-a-custom-field) or a built in component. [See here](#list-of-field-components) for a full list of custom components.

> It is considered a good practiced to have your plugins in a separate file, this allows the plugin to be lazy loaded only when the CMS is enabled. This way it does not effect your production bundle. 

Reviewer: ```suggestion
It is considered a good practice to have your plugins in a separate file, this allows the plugin to be lazy-loaded only when the CMS is enabled. This way it does not affect your production bundle. 
```

```tsx
// ./plugins.tsx
import { TextField } from 'tinacms'
 
export const validationPlugin = {
    Component: TextField,
    name: "text-email-validation",
    validate: (email, allValues, meta, field)=> {
       let isValidEmail = /.*@.*\..*/.test(email)
       if (!isValidEmail) return 'Invalid email address'
    },
}
```

### 2. Register a Field Plugin

> The plugin can then be registered in [the CMS callback](https://tina.io/docs/tinacms-context/#tinacms) in the `<Tina>` wrapper component.

Reviewer: Can it be registered elsewhere?

Reviewer: ```suggestion
The plugin can then be registered in [the CMS callback](/docs/tinacms-context/#tinacms) in the `<Tina>` wrapper component.
```

```tsx
cmsCallback={cms => {
    import('../plugins.tsx').then(({validationPlugin})=>{
        cms.plugins.add(validationPlugin)
    })
}}
```

### 3. Use Field in  `.tina/schema.ts`

> Now in the [schema.ts file](https://tina.io/docs/schema/) this field can be used for any field. It can be added to the [`ui` property](https://tina.io/docs/schema/#the-ui-property)

Reviewer: ```suggestion
Now in the [schema.ts file](/docs/schema/) this field can be used for any field. It can be added to the [`ui` property](/docs/schema/#the-ui-property)
```
